### PR TITLE
CFE-4244: Add the function name to the result cache key

### PR DESCRIPTION
**Pull request by amousset**
The function cache only used the args values as key, which in some cases could lead to mixing results from different functions with the same arguments.

Related to #5313.

### CODE REVIEW

File: libpromises/eval_context.c [hunk 1]
```diff
@@ -2863,7 +2866,11 @@
 
     Rval *rval_copy = xmalloc(sizeof(Rval));
     *rval_copy = RvalCopy(*rval);
-    FuncCacheMapInsert(ctx->function_cache, RlistCopy(args), rval_copy);
+
+    Rlist *args_copy = RlistCopy(args);
+    Rlist *key = RlistPrepend(&args_copy, fp->name, RVAL_TYPE_SCALAR);
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libpromises/eval_context.c:2874]: @amousset you didn't really introduce these errors, but maybe you can fix them while here?

Should be fixed by just adding this to the beginning of the function:

```
assert(fp != NULL);
assert(ctx != NULL);
```

**Comment by amousset** [comment]
The acceptance test fails on Windows:

```
./01_vars/02_functions/cache_name.cf FAIL (UNEXPECTED FAILURE)
```

Maybe a difference in the way dns resolution is implemented. I will try to find a safer test case.

File: tests/acceptance/01_vars/02_functions/cache_name.cf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#######################################################
+
+bundle agent test
+{
+  meta:
+    "description" -> { "CFE-4244" }
+      string => "Test that the function result cache checks function name";
+
+  vars:
+    "res1" string => format("cfengine.com");
```
Comment (amousset) [inline tests/acceptance/01_vars/02_functions/cache_name.cf:30]: I think these functions do not test the actual problem as they are not cached (only those with the `FNCALL_OPTION_CACHED` flag are IIRC).

**Comment by larsewi** [comment]
Tried a third attempt to create an acceptance test. Let me know what you think @amousset 
```
----------------------------------------------------------------------
./01_vars/02_functions/cache_name.cf 
----------------------------------------------------------------------
R: test description: Test that the function result cache checks function name
R: /home/vagrant/core/tests/acceptance/./01_vars/02_functions/cache_name.cf Pass
R: res1!=res2 -> 'cfengine.com'!='cfengine_com'

Return code is 0.

  ==> Pass
```

File: tests/acceptance/01_vars/02_functions/cache_name.cf (hunk 2 above)
Comment (larsewi) [reply tests/acceptance/01_vars/02_functions/cache_name.cf:30]: Ahaa, okey.

**Comment by amousset** [comment]
Looks good :+1:

**Comment by amousset** [comment]
> Please ACK if you agree with the results @amousset :)

yes :+1:

**Comment by larsewi** [comment]
@cf-bottom can you run this through Jenkins please?

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9818/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9818/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9818/

**Review by larsewi — APPROVED** [review]
Thanks! Let's wait for CI to approve

**Review by vpodzime — APPROVED** [review]
ACK

**Comment by larsewi** [comment]
Thanks for your contribution @amousset 🚀

**Pull request merged**